### PR TITLE
Interface refactor

### PR DESCRIPTION
**Pull request by DinoBektesevic**
Draft PR for discussions on proposed refactoring

### CODE REVIEW

**Review by DinoBektesevic — COMMENTED** [review]
I like modularization and groupings of the code in the proposal, but my problem is that I can write some C++ code but I've never written a library in it so I'm not familiar with all the implicit tradeoffs and best practices.

File: src/kbmod/image_base/bindings.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+namespace py = pybind11;
+
+using pf = image_base::PointSpreadFunc;
+using ri = image_base::RawImage;
+using li = image_base::LayeredImage;
+using is = image_base::ImageStack;
+
+using std::to_string;
+
+PYBIND11_MODULE(search, m) {
```
[the remark this answers is not in the document]
Comment (jeremykubica) [reply src/kbmod/image_base/bindings.cpp:20]: Interesting. I didn't realize that was a restriction. We should chat about how we can break things up then.

**Comment by DinoBektesevic** [comment]
I just pushed a proposal example on the PSF class of how to split the implementation up into multiple functions. I believe the C++ code can be made to be self-standing in this way as well (i.e. it can be compiled independently of the presence of Pybind with some modifications to the CMake script). I have no idea if that's useful at all. 

I saw that you wanted to split up the modules via multiple `PYBIND11_MODULE` macros, but I'm just not sure that that is possible. I tried at least renaming the module name of the one, because I'm certain extending an already declared module is not possible, but it doesn't seem to work either. See for example:

```python
>>> import kbmod.search
>>> dir(kbmod.search)
['STAMP_MEAN', 'STAMP_MEDIAN', 'STAMP_SUM', 'StampType', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'calculate_likelihood_psi_phi', 'sigmag_filtered_indices', 'stack_search', 'stamp_parameters', 'trajectory']

>>> import kbmod.image_base
>>> dir(kbmod.image_base)
['__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__']
>>> 
```
There's just nothing in the module at all. The proposed workaround is to have a bindings code, including the pythonic docstrings, for each of the classes we interface to Python instead. 

One thing I don't like is that the factories, because the implementation is in the "header" gets duplicated everywhere so I've declared all as "static" I don't know if that's the best approach, or what the tradeoff is, but how the code gets unraveled by the linker is somewhat of a mystery to me.

**Comment by jeremykubica** [comment]
Closing as we decided not to go down this path at the moment and Dino already implemented the binding refactor in #346

**Pull request closed without merging**